File: Serenity.Script.UI/PropertyDialog/PropertyPanel.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [IncludeGenericArguments(false), ScriptName("PropertyPanel")]
    public abstract partial class PropertyPanel<TEntity, TOptions> : TemplatedPanel<TOptions>
        where TEntity : class, new()
        where TOptions: class, new()
    {
        private TEntity entity;
        private Int64? entityId;

        protected PropertyPanel(jQueryObject div, TOptions opt = null)
            : base(div, opt)
        {
            if (!IsAsyncWidget())
            {
                #pragma warning disable 618
                InitPropertyGrid();
                LoadInitialEntity();
                #pragma warning restore 618
            }
        }

        protected override Promise InitializeAsync()
        {
            return base.InitializeAsync()
                .ThenAwait(InitPropertyGridAsync)
                .Then(() =>
                {
                    LoadInitialEntity();
                });
        }

        protected virtual void LoadInitialEntity()
        {
            if (propertyGrid != null)
                propertyGrid.Load(new object());
        }

        public override void Destroy()
        {
            if (propertyGrid != null)
            {
                propertyGrid.Destroy();
                propertyGrid = null;
            }

            if (validator != null)
            {
                this.ById("Form").Remove();
                validator = null;
            }

            base.Destroy();
        }

        protected TEntity Entity
        {
            get { return entity; }
            set { entity = value ?? new object().As<TEntity>(); }
        }

        protected internal Nullable<Int64> EntityId
        {
            get { return entityId; }
            set { entityId = value; }
        }

        protected PropertyGrid propertyGrid;

        private void InitPropertyGrid()
        {
            var pgDiv = this.ById("PropertyGrid");
            if (pgDiv.Length <= 0)
                return;

            #pragma warning disable 618
            var pgOptions = GetPropertyGridOptions();
            #pragma warning restore 618

            propertyGrid = new PropertyGrid(pgDiv, pgOptions).Init();
            if (this.element.Closest(".ui-Panel").HasClass("s-Flexify"))
                propertyGrid.Element.Children(".categories").FlexHeightOnly();
        }

        private Promise InitPropertyGridAsync()
        {
            return Promise.Void.ThenAwait(() =>
            {
                var pgDiv = this.ById("PropertyGrid");
                if (pgDiv.Length <= 0)
                    return Promise.Void;

                return GetPropertyGridOptionsAsync()
                    .ThenAwait(pgOptions =>
                    {
                        propertyGrid = new PropertyGrid(pgDiv, pgOptions);
                        if (this.element.Closest(".ui-Panel").HasClass("s-Flexify"))
                            propertyGrid.Element.Children(".categories").FlexHeightOnly();
                        return propertyGrid.Initialize();
                    });
            });
        }

        protected virtual string GetFormKey()
        {
            var attributes = this.GetType().GetCustomAttributes(typeof(FormKeyAttribute), true);
            if (attributes.Length >= 1)
                return attributes[0].As<FormKeyAttribute>().Value;

            var name = this.GetType().FullName;
            var px = name.IndexOf(".");
            if (px >= 0)
                name = name.Substring(px + 1);

            if (name.EndsWith("Panel"))
                name = name.Substr(0, name.Length - 6);
            else if (name.EndsWith("Panel"))
                name = name.Substr(0, name.Length - 5);

            return name;
        }

        protected virtual List<PropertyItem> GetPropertyItems()
        {
            #pragma warning disable 618
            var formKey = GetFormKey();
            return Q.GetForm(formKey);
            #pragma warning restore 618
        }

        protected virtual Promise<List<PropertyItem>> GetPropertyItemsAsync()
        {
            return Promise.Void.ThenAwait(() => 
            {
                var formKey = GetFormKey();
                return Q.GetFormAsync(formKey);
            });
        }

        protected virtual PropertyGridOptions GetPropertyGridOptions()
        {
            #pragma warning disable 618
            return new PropertyGridOptions
            {
                IdPrefix = this.idPrefix,
                Items = GetPropertyItems(),
                Mode = PropertyGridMode.Insert,
                UseCategories = false,
                LocalTextPrefix = "Forms." + GetFormKey() + "."
            };
            #pragma warning restore 618
        }

        protected virtual Promise<PropertyGridOptions> GetPropertyGridOptionsAsync()
        {
            return GetPropertyItemsAsync().ThenSelect(propertyItems =>
            {
                return new PropertyGridOptions
                {
                    IdPrefix = this.idPrefix,
                    Items = propertyItems,
                    Mode = PropertyGridMode.Insert,
                    UseCategories = false,
                    LocalTextPrefix = "Forms." + GetFormKey() + ".",
                };
            });
        }

        protected virtual bool ValidateBeforeSave()
        {
            return this.validator.ValidateForm();
        }

        protected virtual TEntity GetSaveEntity()
        {
            var entity = new object().As<TEntity>();

            if (this.propertyGrid != null)
                this.propertyGrid.Save(entity);

            return entity;
        }
    }

    [Imported, IncludeGenericArguments(false), ScriptName("PropertyPanel")]
    public abstract partial class PropertyPanel<TEntity> : PropertyPanel<TEntity, object>
        where TEntity : class, new()
    {
        public PropertyPanel(jQueryObject container)
            : base(container, null)
        {
        }
    }
}

File: Serenity.Scripts/CoreLib/UI/Widgets/PropertyPanel.ts
namespace Serenity {

    @Decorators.registerClass('Serenity.PropertyPanel')
    export class PropertyPanel<TItem, TOptions> extends TemplatedPanel<TOptions> {

        private _entity: TItem;
        private _entityId: any;

        constructor(container: JQuery, options?: TOptions) {
            super(container, options);

            this.initPropertyGrid();
            this.loadInitialEntity();
        }

        destroy() {
            if (this.propertyGrid) {
                this.propertyGrid.destroy();
                this.propertyGrid = null;
            }
            if (this.validator) {
                this.byId('Form').remove();
                this.validator = null;
            }
            super.destroy();
        }

        protected initPropertyGrid() {
            var pgDiv = this.byId('PropertyGrid');
            if (pgDiv.length <= 0) {
                return;
            }
            var pgOptions = this.getPropertyGridOptions();
            this.propertyGrid = (new Serenity.PropertyGrid(pgDiv, pgOptions)).init(null);
            if (this.element.closest('.ui-Panel').hasClass('s-Flexify')) {
                this.propertyGrid.element.children('.categories').flexHeightOnly(1);
            }
        }

        protected loadInitialEntity(): void {
            if (this.propertyGrid) {
                this.propertyGrid.load(new Object());
            }
        }

        protected getFormKey(): string {
            var attributes = Q.getAttributes(
                Q.getInstanceType(this), Serenity.FormKeyAttribute, true);

            if (attributes.length >= 1) {
                return attributes[0].value;
            }
            var name = Q.getTypeFullName(Q.getInstanceType(this));
            var px = name.indexOf('.');
            if (px >= 0) {
                name = name.substring(px + 1);
            }
            if (Q.endsWith(name, 'Panel')) {
                name = name.substr(0, name.length - 6);
            }
            else if (Q.endsWith(name, 'Panel')) {
                name = name.substr(0, name.length - 5);
            }
            return name;
        }

        protected getPropertyGridOptions(): PropertyGridOptions {
            return {
                idPrefix: this.idPrefix,
                items: this.getPropertyItems(),
                mode: 1,
                useCategories: false,
                localTextPrefix: 'Forms.' + this.getFormKey() + '.'
            };
        }

        protected getPropertyItems(): PropertyItem[] {
            var formKey = this.getFormKey();
            return Q.getForm(formKey);
        }

        protected getSaveEntity(): TItem {
            var entity = new Object();
            if (this.propertyGrid) {
                this.propertyGrid.save(entity);
            }
            return entity as TItem;
        }

        protected get_entity(): TItem {
            return this._entity;
        }

        protected get_entityId(): any {
            return this._entityId;
        }
        
        protected set_entity(value: TItem): void {
            this._entity = Q.coalesce(value, new Object());
        }

        protected set_entityId(value: any): void {
            this._entityId = value;
        }

        protected validateBeforeSave(): boolean {
            return this.validator.form();
        }

        protected propertyGrid: Serenity.PropertyGrid;
    }
}